### PR TITLE
Add OGR_G_Normalize() and bind it to SWIG

### PR DESCRIPTION
**Pull request by kmuehlbauer**
## What does this PR do?

This PR adds GEOSNormalize functionality to OGR as OGR_G_Normalize() and binds it to SWIG 

## Tasklist

 - [x] Add OGR_G_Normalize() and bind it to SWIG 
 - [x] Add test case(s)
 - [ ] Add documentation
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
Hmm, should I add the generated Swig code to the PR?

**Comment by rouault** [comment]
> Hmm, should I add the generated Swig code to the PR?

yes, please, and with SWIG 3.0.12 (can be reasonably be easily built from source: https://sourceforge.net/projects/swig/files/swig/swig-3.0.12/)

**Comment by kmuehlbauer** [comment]
@rouault Thanks for the advice. Not sure, why the android and mingw builds are failing. Please let me know, if there is anything left for me to do.

File: gdal/ogr/ogrgeometry.cpp [hunk 1]
```diff
@@ -3693,6 +3693,83 @@ OGRGeometryH OGR_G_MakeValid( OGRGeometryH hGeom )
...
+/************************************************************************/
+/*                         OGR_G_Normalize()                            */
+/************************************************************************/
+
+/**
+ * \brief Attempts to bring geometry into normalized/canonical form.
+ *
+ * This function is the same as the C++ method OGRGeometry::Normalize().
+ *
+ * This function is built on the GEOS >= 3.8 library, check it for the definition
```
Comment (dbaston) [inline gdal/ogr/ogrgeometry.cpp:3753]:
```suggestion
 * This function is built on the GEOS library; check it for the definition
```

File: gdal/ogr/ogrgeometry.cpp [hunk 2]
```diff
@@ -3693,6 +3693,83 @@ OGRGeometryH OGR_G_MakeValid( OGRGeometryH hGeom )
...
+ * of the geometry operation.
+ * If OGR is built without the GEOS library, this function will always fail,
+ * issuing a CPLE_NotSupported error.*
```
Comment (dbaston) [inline gdal/ogr/ogrgeometry.cpp:3756]:
```suggestion
 * issuing a CPLE_NotSupported error.
```

File: gdal/ogr/ogrgeometry.cpp [hunk 3]
```diff
@@ -3693,6 +3693,83 @@ OGRGeometryH OGR_G_MakeValid( OGRGeometryH hGeom )
...
+/************************************************************************/
+/*                            Normalize()                               */
+/************************************************************************/
+
+/**
+ * \brief Attempts to bring geometry into normalized/canonical form.
+ *
+ * This method is the same as the C function OGR_G_Normalize().
+ *
+ * This function is built on the GEOS >= 3.8 library, check it for the definition
```
Comment (dbaston) [inline gdal/ogr/ogrgeometry.cpp:3705]:
```suggestion
 * This function is built on the GEOS library; check it for the definition
```

File: gdal/ogr/ogrgeometry.cpp [hunk 4]
```diff
@@ -3693,6 +3693,83 @@ OGRGeometryH OGR_G_MakeValid( OGRGeometryH hGeom )
...
+    {
+
+        int hGEOSRet = GEOSNormalize_r( hGEOSCtxt, hGeosGeom );
+
+        if( hGEOSRet == 0 )
+        {
+             poOGRProduct = BuildGeometryFromGEOS(hGEOSCtxt, hGeosGeom,
+                                                 this, nullptr);
+
+        }
```
Comment (dbaston) [inline gdal/ogr/ogrgeometry.cpp:3736]: Not sure why `GEOSNormalize_r` would ever fail, but if it did, we would leak here.

File: gdal/ogr/ogrgeometry.cpp [hunk 5]
```diff
@@ -3693,6 +3693,83 @@ OGRGeometryH OGR_G_MakeValid( OGRGeometryH hGeom )
...
+ * on failure.
+ *
+ * @since GDAL 3.3
+ */
+OGRGeometry *OGRGeometry::Normalize() const
+{
+#ifndef HAVE_GEOS
+    CPLError( CE_Failure, CPLE_NotSupported,
+                  "GEOS support not enabled." );
+        return -1.0;
```
Comment (dbaston) [inline gdal/ogr/ogrgeometry.cpp:3720]:
```suggestion
        return nullptr;
```

File: gdal/ogr/ogrgeometry.cpp (hunk 4 above)
Comment (kmuehlbauer) [reply gdal/ogr/ogrgeometry.cpp:3736]: @dbaston What could be done prevent leakage? Add an else with cleanup of the context and returning some error message?

File: gdal/ogr/ogrgeometry.cpp [hunk 6]
```diff
@@ -3693,6 +3693,83 @@ OGRGeometryH OGR_G_MakeValid( OGRGeometryH hGeom )
...
+
+        int hGEOSRet = GEOSNormalize_r( hGEOSCtxt, hGeosGeom );
+
+        if( hGEOSRet == 0 )
+        {
+             poOGRProduct = BuildGeometryFromGEOS(hGEOSCtxt, hGeosGeom,
+                                                 this, nullptr);
+
+        }
+    }
```
Comment (dbaston) [inline gdal/ogr/ogrgeometry.cpp:3737]:
```suggestion
    } else {
              GEOSGeom_destroy_r( hGEOSCtxt, hGeosGeom );
    }
```

File: gdal/ogr/ogrgeometry.cpp (hunk 4 above)
Comment (kmuehlbauer) [reply gdal/ogr/ogrgeometry.cpp:3736]: Thanks for the suggestion @dbaston!

File: gdal/ogr/ogrgeometry.cpp [hunk 7]
```diff
@@ -3693,6 +3693,85 @@ OGRGeometryH OGR_G_MakeValid( OGRGeometryH hGeom )
...
+        int hGEOSRet = GEOSNormalize_r( hGEOSCtxt, hGeosGeom );
+
+        if( hGEOSRet == 0 )
+        {
+             poOGRProduct = BuildGeometryFromGEOS(hGEOSCtxt, hGeosGeom,
+                                                 this, nullptr);
+
+        }
+    } else {
+              GEOSGeom_destroy_r( hGEOSCtxt, hGeosGeom );
```
Comment (rouault) [inline gdal/ogr/ogrgeometry.cpp:3738]: this should be moved one level deeper, as the else clause of ``` if( hGEOSRet == 0 )```

**Review by rouault — CHANGES_REQUESTED** [review]

**Review by rouault — APPROVED** [review]

**Pull request merged**